intermediate_source/custom_function_conv_bn_tutorial.py

```diff
@@ -0,0 +1,344 @@
+    # Helper function to unsqueeze the dimensions that we reduce over
+    return t[None, :, None, None]
+
+def batch_norm_backward(grad_out, X, sum, sqrt_var, N, eps):
+    # To simplify our derivation, we follow the chain rule and compute
+    # the following gradients before accumulating them all into a final grad_input
+    #  1) X of the numerator
+    #  2) mean of the numerator
+    #  3) var of the denominator
+    # d_denom = -num / denom**2
+    tmp = -((X - expand(sum) / N) * grad_out).sum(dim=(0, 2, 3))
+    d_denom = tmp / (sqrt_var + eps)**2
+    # It is useful to delete tensors when you no longer need them
+    # It's not a big difference here though because tmp only has size of (C,)
+    # The important thing is avoid allocating NCHW-sized tensors unnecessarily
+    del tmp
+    grad_input = grad_out / expand(sqrt_var + eps)  # = d_numerator
+    grad_input -= expand(grad_input.sum(dim=(0, 2, 3))) / N  # mean = X.sum(dim=(0, 2, 3)) / N
+    # denom = torch.sqrt(var) + eps
+    d_var = d_denom / (2 * sqrt_var)
+    del d_denom
+    # unbiased_var(x) = ((X - expand(mean))**2).sum(dim=(0, 2, 3)) / (N - 1)
+    grad_input += (X * expand(d_var) * N + expand(-d_var * sum)) * 2 / ((N - 1) * N)
+    return grad_input
+
+class BatchNorm(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, X, eps=1e-6):
+        # Don't save keepdim'd values for backward
+        sum = X.sum(dim=(0, 2, 3))
+        var = X.var(unbiased=True, dim=(0, 2, 3))
+        N = X.numel() / X.size(1)
+        sqrt_var = torch.sqrt(var)
+        ctx.save_for_backward(X)
+        ctx.eps = eps
+        ctx.sum = sum
+        ctx.N = N
+        ctx.sqrt_var = sqrt_var
+        mean = sum / N
+        denom = sqrt_var + eps
+        return (X - expand(mean)) / expand(denom)
+
+    @staticmethod
+    @once_differentiable
+    def backward(ctx, grad_out):
+        X, = ctx.saved_tensors
+        return batch_norm_backward(grad_out, X, ctx.sum, ctx.sqrt_var, ctx.N, ctx.eps)
+
+######################################################################
+# Testing with gradcheck
+a = torch.rand(1, 2, 3, 4, requires_grad=True, dtype=torch.double)
+torch.autograd.gradcheck(BatchNorm.apply, (a,), fast_mode=False)
+
+######################################################################
+# Now that the bulk of the work has been done, we can combine
+# them together. As seen in (1) of the forward pass below, we only need to
+# save a single buffer for backward, but we need to recompute
+# Conv2D again in (2).
+class FusedConvBN2DFunction(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, X, conv_weight, running_mean, running_var, training, exp_avg_factor=0.1, eps=1e-6):
+        assert not running_mean.requires_grad and not running_var.requires_grad
+        assert X.ndim == 4
+        ctx.save_for_backward(X, conv_weight)  # (1) Only need to save this single buffer for backward!
+        X = F.conv2d(X, conv_weight)
+
+        if training:
```

Comment:
Why handle this here if you don't in the previous example?

Reply:
It would be simpler to just not handle the non-training case to simply fall back to the example above.

Reply:
This would be a big simplification indeed. Its probably fine since we can do batched testing here.